docs/docs/server-middleware.md

```diff
@@ -0,0 +1,649 @@
+# Server Middleware
+
+Http4s includes some middleware out of the box in the `org.http4s.server.middleware`
+package. Some of it is documented in its own page:
+
+* [Authentication]
+* Cross Origin Resource Sharing ([CORS])
+* Response Compression ([GZip])
+* [HSTS]
+* [CSRF]
+
+We'll describe and provide examples for the remaining middleware, but first we set up our service:
+```scala mdoc:silent
+import cats.effect._
+import cats.syntax.all._
+import org.typelevel.ci._
+import org.http4s._
+import org.http4s.dsl.io._
+import org.http4s.implicits._
+import cats.effect.unsafe.IORuntime
+import scala.concurrent.duration._
+import cats.effect.std.Random
+import fs2.Stream
+
+implicit val runtime: IORuntime = cats.effect.unsafe.IORuntime.global
+
+object NameQueryParamMatcher extends QueryParamDecoderMatcher[String]("name")
+
+val service = HttpRoutes.of[IO] {
+  case GET -> Root / "bad" => BadRequest()
+  case GET -> Root / "ok" => Ok()
+  case r @ POST -> Root / "post" => r.as[Unit] >> Ok()
+  case GET -> Root / "b" / "c" => Ok()
+  case POST -> Root / "queryForm" :? NameQueryParamMatcher(name) => Ok(s"hello $name")
+  case GET -> Root / "wait" => IO.sleep(10.millis) >> Ok()
+  case GET -> Root / "boom" => IO.raiseError(new RuntimeException("boom!"))
+  case r @ GET -> Root / "reverse" => r.as[String].flatMap(s => Ok(s.reverse))
```

Comment:
Should this be a POST? Since a GET should not have a body

Reply:
Good point

Reply:
Fixed